www_src/pages/tinker/tinker.jsx

```diff
@@ -1,68 +1,157 @@
+
+/**
+ * Tinker: Advanced editor for specific element attributes or styles
+ *
+ * Props
+ *   @key {boolean} isVisible Indicates if the Tinker Activity is currently visible (not paused)
+
+ * State
+ *   @key {boolean} loading If true, loading spinner is shown and UI is blocked
+ *   @key {object} element Represents flattened element data (via spec.flatten)
 
+ *   From router:
+ *   @key {object} params Contains params user, project, page, element, propertyName
+ */
 var Tinker = React.createClass({
+
+  mixins: [require('../../lib/router')],
+
   getInitialState: function () {
     return {
-      color: '#8FB70A'
+      loading: false,
+      element: null
     };
   },
-  onChangeColor: function (color) {
+
+  componentDidMount: function () {
+
+    // Prevent pull to refresh
+    document.body.style.overflowY = 'hidden';
+
+    this.load();
+
+    this.props.update({
+      onBackPressed: () => {
+        this.save(() => window.Android.goBack());
+      }
+    });
+
+  },
+
+  /**
+   * If tinker activity switches from paused to resumed, re-load data from the API
+   */
+  componentDidUpdate: function (prevProps, prevState) {
+    if (this.props.isVisible && !prevProps.isVisible) {
+      this.load();
+    }
+  },
+
+  /**
+   * Load element data from API, based on router params
+   */
+  load: function () {
+    this.setState({loading: true});
+    api.getElement(this.state.params, (err, element) => {
+      this.setState({loading: false});
+      if (err) {
+        return console.error(err);
+      }
+      var spec = elementTypes[element.type].spec;
+      if (!spec.spec[this.state.params.propertyName]) {
+        return;
+      }
+
+      this.setState({
+        element: spec.flatten(element),
+        editor: spec.spec[this.state.params.propertyName].editor || 'color'
+      });
+    });
+  },
+
+  /**
+   * Save current element state in API
+   * Triggered in componentDidUpdate
+   */
+  save: function (onSaveComplete) {
+    var element = this.state.element;
+    var propertyName = this.state.params.propertyName;
+    var spec = elementTypes[element.type].spec;
+    var isStyleOrAttribute = spec.isStyleOrAttribute(propertyName);
+    var updateObject = {};
+    updateObject[isStyleOrAttribute] = spec.expand(element)[isStyleOrAttribute];
+
+    this.setState({loading: true});
+    var options = assign({}, this.state.params, {json: updateObject});
+    api.updateElement(options, (err, element) => {
+      this.setState({loading: false});
+      if (err) {
+        console.error(err);
+      }
+      if (typeof onSaveComplete === 'function') {
+        onSaveComplete();
+      }
+    });
+  },
+
+  getEditorValue: function () {
+    return this.state.element && this.state.element[this.state.params.propertyName];
+  },
+
+  setEditorValue: function (value) {
+    var edits = {};
+    edits[this.state.params.propertyName] = value;
     this.setState({
-      color: color
+      element: assign(this.state.element, edits)
```

Comment:
is this safe, or should this be `var thing = assign({}, this.state.element)` + `assign(thing, edits)` instead, so that we don't modify `this.state.element`?